### PR TITLE
Merge Mip 69

### LINKED ISSUE #69: [Technical Improvement] Update Password Encoder Algorithm
## Rationale
Our password encoder algorithm is outdated and not secure (SHA-256)
We need to update to a more secure one

In addition, the usage of RememberMe token is not secure.

## 1. Functional Requirements
No functionnal modification 

### Security 
We want to use the Argon2id algorithm to hash passwords with parameters 
·	int opsLimit = 2; // 2 iterations
·	int memLimit = 32768; // 32 MB
·	int parallelism = 1; // 1 parallel execution
·	int outputLength = 32; // password hash length
The password Salt has to be 128 bits (16 bytes) of length.

For RememberMe Token, we want to generate it with the Argon2id algorithm with same parameters as password

### Configurability 
As Argon2 parameters could not be changed without reseting all password, we not make it configurable

### Upgradability
The migration strategy have 3 options : 
- keep hash with old algo. When user logs in, update the password hash with the new algo
- Invalidate passwords, so thats users have to update password, which will be hashed with new algo. Not applicable in our case as we cannot "invalidate" password
- Update SHA-256 hashed passwords stored in database. Database stores as password hash oldAlgo(password). Upgrade plugin will update theses hashes to have newAlgo(oldAlgo(password)). So that, all password are hashed with the new algo. The login processus must be able to validate password in both options.

We will use the most secure strategy based on data upgrade : the solution 3

### Existing Features
No modification

### Feature Flags
No Feature flag

### Documentation 
Technical documentation to update

## 4. Software Architecture
Store password process : 
image

Store Password Sequence Diagramm
image

Only the red part is to create

Validate Password process
image

Validate Password pseudo code in AttributeManager.validatePassword (afeter data upgrade execution)
```
Input : username, password

If (not user have attribute passwordSalt128 && user have attribute passwordSalt) {
   // user not connected && not migrated by data upgrade
   Check password with old algo (oldAlgo(password) == hashInDatabase ?)
   If validation is ok, call updatePassword to compute and save a hash with the new algo. During passwordUpdate, we need to ensure to remove passwordSalt attribute

} else if(user have attribute passwordSalt128 && user have attribute passwordSalt) (
   // user not connected && migrated by data upgrade
   Hash = Compute the hash of the password by using the old passwordEncoder  DatabaseReadingSaltEncoder. (oldAlgo(password))
   Then check hash with newAlgo (newAlgo(oldAlgo(password)) == hashInDatabase ?)
   If validation is ok, call updatePassword to compute and save a hash with the new algo. During passwordUpdate, we need to ensure to remove passwordSalt attribute

} else {
  //user connected
  use the current way to check password it will get the default configured PasswordEncoder and use it
      return this.getRepository().validateCredential(this.getInvocationContext(), this.createIdentityObject(user), new PasswordCredential(hash, this.getCredentialEncoder(), user.getKey()));
}

```

Remember me generation : 
image

RememberMe Validation
image

Once remember me token are generated with new algo, we need an AuthenticatorPlugin to authenticate the user.

### Access 
   - GUI 
   - API 

### Services & processing
   - Ex: Scheduled Jobs

### Data and persistence
   - DataUpgrade 

```
for all user without attribute “passwordSalt128” :
   Get password hash in database
   Generate a passwordSalt128 and store it as attribute for the user
   Compute newAlgo(hash) and store it as new hash for the user
```
## 5. Annexes
   - Reasoning that explain the technical choices
   - Impacts 
   - References

### PR DESCRIPTION
**Pull request by rdenarie**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -22,12 +22,12 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
   <parent>
     <artifactId>addons-parent-pom</artifactId>
     <groupId>org.exoplatform.addons</groupId>
-    <version>17-M04</version>
+    <version>17-mips-SNAPSHOT</version>
```
Comment (boubaker) [inline pom.xml:25]:
```suggestion
    <version>17-M04</version>
```

File: pom.xml [hunk 2]
```diff
@@ -22,12 +22,12 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
     <relativePath/>
   </parent>
   <groupId>org.exoplatform.addons.wallet</groupId>
   <artifactId>wallet</artifactId>
-  <version>2.5.x-SNAPSHOT</version>
+  <version>2.5.x-mips-SNAPSHOT</version>
```
Comment (boubaker) [inline pom.xml:30]:
```suggestion
  <version>2.5.x-SNAPSHOT</version>
```

File: pom.xml [hunk 3]
```diff
@@ -46,10 +46,10 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
   </scm>
   <properties>
     <!-- 3rd party libraries versions -->
-    <org.exoplatform.social.version>6.5.x-SNAPSHOT</org.exoplatform.social.version>
+    <org.exoplatform.social.version>6.5.x-mips-SNAPSHOT</org.exoplatform.social.version>
```
Comment (boubaker) [inline pom.xml:49]:
```suggestion
    <org.exoplatform.social.version>6.5.x-SNAPSHOT</org.exoplatform.social.version>
```

File: pom.xml [hunk 4]
```diff
@@ -46,10 +46,10 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
     <org.exoplatform.wallet.ert-contract.version>1.3.0</org.exoplatform.wallet.ert-contract.version>
-    <org.exoplatform.platform-ui.version>6.5.x-SNAPSHOT</org.exoplatform.platform-ui.version>
-    <addon.exo.analytics.version>1.4.x-SNAPSHOT</addon.exo.analytics.version>
+    <org.exoplatform.platform-ui.version>6.5.x-mips-SNAPSHOT</org.exoplatform.platform-ui.version>
+    <addon.exo.analytics.version>1.4.x-mips-SNAPSHOT</addon.exo.analytics.version>
```
Comment (boubaker) [inline pom.xml:52]:
```suggestion
    <org.exoplatform.platform-ui.version>6.5.x-SNAPSHOT</org.exoplatform.platform-ui.version>
    <addon.exo.analytics.version>1.4.x-SNAPSHOT</addon.exo.analytics.version>
```

**Review by boubaker — COMMENTED** [review]
Please delete MIPs FB version change commit

**Comment by rdenarie** [comment]
I close this pr as it is not the correct source branch

**Pull request closed without merging**